File: src/main/java/org/pl/interpreter/visitor/ProgramVisitor.java
package org.pl.interpreter.visitor;

import org.pl.interpreter.IInterpreter;
import org.pl.interpreter.exception.VisitorException;
import org.pl.parser.ast.INode;
import org.pl.parser.ast.ProgramNode;

/**
 * Interprets the program logic.
 */
public class ProgramVisitor implements IVisitor {

    @Override
    public boolean matches(INode node) {
        return ProgramNode.class == node.getClass();
    }

    @Override
    public Object visit(IInterpreter interpreter, INode node) throws VisitorException {
        var programNode = (ProgramNode) node;
        return interpreter.evalNode(programNode.getStatements());
    }
}

File: src/main/kotlin/org/leaflang/interpreter/visitor/ProgramVisitor.kt
package org.leaflang.interpreter.visitor

import org.leaflang.interpreter.IInterpreter
import org.leaflang.interpreter.result.IRuntimeResult
import org.leaflang.parser.ast.INode
import org.leaflang.parser.ast.ProgramNode

/**
 * Interprets the program logic.
 */
class ProgramVisitor : IVisitor {
    override fun visit(interpreter: IInterpreter, node: INode): IRuntimeResult {
        val programNode = node as ProgramNode
        return interpreter.interpret(programNode.statements)
    }
}